crates/wasmtime/src/runtime/ref/gc_ref/rooting.rs

```diff
@@ -0,0 +1,1423 @@
+    }
+
+    fn is_lifo(&self) -> bool {
+        self.discriminant() == Self::LIFO_DISCRIMINANT
+    }
+
+    fn is_manual(&self) -> bool {
+        self.discriminant() == Self::MANUAL_DISCRIMINANT
+    }
+
+    fn payload(&self) -> u32 {
+        self.0 & Self::PAYLOAD_MASK
+    }
+
+    fn as_lifo(&self) -> Option<usize> {
+        if self.is_lifo() {
+            Some(usize::try_from(self.payload()).unwrap())
+        } else {
+            None
+        }
+    }
+
+    fn as_manual(&self) -> Option<SlabId> {
+        if self.is_manual() {
+            Some(SlabId::from_raw(self.payload()))
+        } else {
+            None
+        }
+    }
+}
+
+/// The set of all GC roots in a single store/heap.
+#[derive(Debug, Default)]
+pub(crate) struct RootSet {
+    /// GC roots with arbitrary lifetime that are manually rooted and unrooted,
+    /// for use with `ManuallyRooted<T>`.
+    manually_rooted: Slab<VMGcRef>,
+
+    /// Strictly LIFO-ordered GC roots, for use with `RootScope` and
+    /// `Rooted<T>`.
+    lifo_roots: Vec<LifoRoot>,
+
+    /// Generation counter for entries to prevent ABA bugs with `RootScope` and
+    /// `Rooted<T>`.
+    lifo_generation: u32,
+}
+
+impl Drop for RootSet {
+    fn drop(&mut self) {
+        // Drop our `ExternRef` roots. In the future this will be a series of
+        // calls to `wasmtime_runtime::GcRuntime::on_unroot` trait method hook
+        // or something like that. However, this is all unnecessary for
+        // non-reference-counting collectors.
+
+        self.exit_lifo_scope(0);
+
+        for (_id, gc_ref) in self.manually_rooted.drain() {
+            // (Inlined copy of `self.unroot_gc_ref(gc_ref)` to avoid borrowing
+            // `self` while `self.manually_rooted` is already borrowed).
+            //
+            // Safety: our mutable access to the root set means that no one else
+            // should have concurrent access to the `VMExternRef`, so
+            // decrementing the reference count here is safe.
+            unsafe {
+                let _ = extern_ref_from_gc_ref(gc_ref);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct LifoRoot {
+    generation: u32,
+    gc_ref: VMGcRef,
+}
+
+impl RootSet {
+    /// Enter a LIFO rooting scope.
+    ///
+    /// Returns an integer that should be passed unmodified to `exit_lifo_scope`
+    /// when the scope is finished.
+    ///
+    /// Calls to `{enter,exit}_lifo_scope` must happen in a strict LIFO order.
+    pub(crate) fn enter_lifo_scope(&self) -> usize {
+        let len = self.lifo_roots.len();
+        log::debug!("Entering GC root set LIFO scope: {len}");
+        len
+    }
+
+    /// Exit a LIFO rooting scope.
+    ///
+    /// The `scope` argument must be the result of the corresponding
+    /// `enter_lifo_scope` call.
+    ///
+    /// Calls to `{enter,exit}_lifo_scope` must happen in a strict LIFO order.
+    pub(crate) fn exit_lifo_scope(&mut self, scope: usize) {
```

Comment:
Since this is on the wasm call hot path, could this be `#[inline]` with a "fast path" of "ok nothing changed" and then have a slow path for "ok gotta `drain`"